### ISSUE TITLE
[2025/06/01] 언어 모델은 왜 이상한 소리를 하는걸까?

### ISSUE THREAD

**Issue by DaewooKim** [DaewooKim/AGI#7]
# [언어 모델은 왜 이상한 소리를 하는걸까?](https://www.linkedin.com/posts/gim-gyungjin-7430aab4_%EC%96%B8%EC%96%B4-%EB%AA%A8%EB%8D%B8%EC%9D%80-%EC%99%9C-%EC%9D%B4%EC%83%81%ED%95%9C-%EC%86%8C%EB%A6%AC%EB%A5%BC-%ED%95%98%EB%8A%94%EA%B1%B8%EA%B9%8C-llm%EC%9D%B4-%EB%84%90%EB%A6%AC-%EC%93%B0%EC%9D%B4%EA%B8%B0-%EC%8B%9C%EC%9E%91%ED%95%98%EB%A9%B4%EC%84%9C-activity-7334560338601746433-haoL?utm_source=share&utm_medium=member_desktop&rcm=ACoAABAfX0cBcPd9w9fX5dTNsijrnYhcRqaSjYc)

LLM이 널리 쓰이기 시작하면서, 모델이 가끔 이상한 소리를 할 수 있다는 점에 대한 사람들의 이해도가 상당히 높아졌다. 이제는 대부분 "모델이 틀릴 수도 있다"는 건 알지만, 막상 본인이 겪는 edge case에서 모델이 원하는 대로 동작하지 않으면, 그냥 모델이 이상하다 느끼는 경우가 많다. 그러나 할루시네이션이라는 현상은 방대한 범위를 담고 있기 때문에 모델을 제대로 평가하고 문제를 해결하려면, 먼저 어떤 유형의 할루시네이션이 존재하는지 파악하고, 내가 겪은 문제가 그중 어떤 유형에 해당하는지 이해할 필요가 있다.

- hallucination 유형[1]
Input-conflicting Hallucination : 사용자가 준 입력 정보나 Instruction과 어긋난 내용을 생성. e.g. (user) "Hill과 나는 농구를 좋아해요", (assistant)"Lucas와 나는 농구를 좋아해요" -> 이름을 잘못됨
Context-conflicting Hallucination : 자신이 이전에 생성한 내용과 모순되는 내용을 생성할 때 발생. 주로 긴 대화나 여러차례의 응답중 기억을 잘 못해서 발생. e.g. 처음엔 "NBA 커미셔너는 Adam Silver"라고 말했다가, 나중엔 "Stern이 그런 말을 했다"고 바꿈 → 전후 모순
Fact-conflicting Hallucination : 세계의 일반 지식(사실)과 충돌하는 정보를 생성할 때 발생 e.g. (user) "포르투갈 왕 Afonso II의 어머니는 누구인가요?" (assistant) "Castile의 Urraca 여왕" (X오답)

2023년만 해도 Fact-conflicting hallucination조차 제대로 해결되지 않아서, 이 부분에 많은 연구가 집중하는 경향이 있었다. 하지만 knowledge conflict은 눈에 띄게 성능이 좋아지면서 사람들은 GPT의 답변에 거의 의문을 품지 않는다. 특히 long context 환경에서의 단순한 input conflict — 예를 들어 needle-in-a-haystack 같은 태스크 — 에 대해서는 거의 해결된 수준이라고 봐도 무방하다[2]. 단순한 QA는 이제 웬만한 base 모델에서 차별점이 되기 힘들다. 요즘은 agent 사용이 늘면서 context가 길어지고, 그에 따라 Input/Context-conflicting hallucination 문제가 더 두드러지고 있는 상황이다.

- hallucination은 왜 일어나는가?
데이터 한계 : LLM 모델은 pretrain 단계에서 여러가지 copus에 대한 기본 지식을 익히고 SFT 단계에서 여러가지 downstream task에 대한 학습을 한다. 이때 pretrain 단계에서 여러가지 충돌되는 지식 copus가 학습되었을 수 있다. Fact check 단계에서는 소수의 신뢰도 있는 문서를 사용하는 것이 성능에 영향을 미치기 때문에 [4] 이를 막기 위해 크롤링 sorce root를 분석하여 공신력 있는 데이터를 선별하는 과정을 거치기도 한다[3].

학습 구조의 한계 : decoding only 방식의 한계. A -> B로 text를 generation하기 때문에 B -> A인 관계를 제대로 학습하지 못할 수 있다. code 문제가 LLM에게 어려운 이유도 코드가 A -> B 순서로 학습할 수 없기 때문이다(import 되는 경우도 있고, 다형성, entity 관계등을 가지고 있기 때문에 언어 모델에서 고치기 쉽지 않음)
entity구조의 데이터를 질의를 하기 위해서는 graph를 많이 사용하는데, 보통 Projector를 만들어 그래프 구조를 학습시킨 다음, 이 projector를 이용해 LLM inference를 하는 식으로 성능을 낸다[6]. 이 과정을 생략할 경우 overfitting이 발생하는데, Input 구조가 text가 아닌 n-dim의 관계일 때는 LLM 자체의 latent space의 케파를 그대로 활용하지 못하는 듯 하다. (output이 일반 언어가 아닌 다른 structure로 변형하는 경우도 유사한 성능 저하가 있는 것으로 보임.[7])

reward hacking : SFT의 가장 큰 문제는 특정 down stream task 에 overfitting 된다는 것이다. 대표적인 문제가 misaligned 문제인데 모델이 token generation 단위로 확률적 사고를 하다보니 중요한 오류와 사소한 오류를 구분하지 않는다. (e.g. 사실을 날조하는 것과 단어 선택의 문제를 동일하게 취급 [5]) 이런 문제는 사용자의 사용성에 상당히 영향을 미치기 때문에 보통 RLHF 학습을 하는데 이 과정에서 reward hacking이 발생한다. A라는 작용을 방지하기 위해 B라는 disadvantage를 주면 그것 disadvantage를 피하기 위해 꼼수 학습을 하는 것이다. (쉬운 예로 부정수급자를 막기 위해 건강보험제도 제도에 여러 제약이 있지만 그 제도의 빈틈을 파고들어 의료쇼핑을 하는 것과 비슷)

- 지금 단계에서 어려운 문제들?
복합적인 entity 관계, 아키텍처 파악이 필요한 이슈 : 특히 코드 문제 같은 경우 상당히 어렵다.
budget 인식이 필요한 이슈 : 모델은 ROI를 따지지며 학습하지 않는다. 그래서 쉬운문제도 과도하게 thinking하다 이상한 답을 내는 경향이 있다(최근엔 이러한 한계가 artificial neural networks 구조 결함 이슈라고 이야기 하기도 함[9]. 인간의 뉴런은 시간 개념을 가지고 있어 입력의 난이도나 복잡성에 따라 연산량을 동적으로 조절가능하지만, 모델은 그렇지 못하는것)
Inductive/deductive 같은 고차원적인 사고 능력이 필요한 이슈 : 대부분 CoT 분석 방식은 top-down 방식. 전략을 정해놓고 답을 택하는 방식이기 때문에, 답은 정해지고 그대로 문제를 풀어나가는 한계가 있다[8]. 문제가 해결되지 않을 경우 새로운 이론을 도입하거나, 단서를 보고 전략을 바꾸는 작업등이 가능해야 고차원 적인 사고가 가능하다

[1] Siren’s Song in the AI Ocean: A Survey on Hallucination in Large Language Models - https://arxiv.org/abs/2309.01219
[2] Listen to a podcast deep dive on long context in Gemini models. https://blog.google/technology/google-deepmind/release-notes-podcast-long-context/
[3] MAmmoTH2 - https://tiger-ai-lab.github.io/MAmmoTH2/
[4] 2024-12 Procedural Knowledge in Pretraining Drives Reasoning in LLMs (Cohear) - https://arxiv.org/abs/2411.12580
[5] 2020-09 Learning to summarize from human feedback(OpenAI) - https://arxiv.org/abs/2009.01325
[6] 2023 GraphGPT: Graph Instruction Tuning for Large Language Models - https://arxiv.org/abs/2310.13023
[7] 2025-05 Don't use structured output mode for reasoning tasks - https://x.com/_weexiao/status/1928150048120131698?s=12
[8] 2025-05 The CoT Encyclopedia: Analyzing, Predicting, and Controlling how a Reasoning Model will Think - https://arxiv.org/abs/2505.10185
[9] 2025-05 Continuous Thought Machines - https://arxiv.org/abs/2505.05522
